backport-2.1: roachtest: use toxiproxy for network testing

tbg opened the pull request:
Backport 1/3 commits from #33282.

Fixes #34136.

/cc @cockroachdb/release

---

Reincarnation of https://github.com/cockroachdb/cockroach/pull/23141.

[toxiproxy] is a TCP proxy which exports an HTTP API that allows
introducing various anomalous network conditions for resilience testing.

Among the interesting things one can simulate are latency and bandwidth
restrictions. This is done programmatically and thus lends itself to the
kind of testing we'd like to do; in particular, it should allow testing
locally a good approximation of geo-replicated clusters and various
network partitions.

The test introduced here isn't really useful yet. We'll need to settle
on something straightforward to test.

Updates #21680.
Updates #14768.

[toxiproxy]: https://github.com/Shopify/toxiproxy

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34144)
<!-- Reviewable:end -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34144#-:-LWl6YBw8Jur3uUnY2p7:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

tbg committed:
client: set timeout on async rollback's context

This fixes a goroutine leak under long-running network partitions.

Release note (bug fix): Fix a goroutine leak that would occur while
a cluster was unavailable (or a subset of nodes partitioned away
from the cluster) and would cause a resource spike to resolve.
